OCPCLOUD-2716: Handle Machine owner references translation between MAPI and CAPI

RadekManak opened the pull request:
This PR adds support to MachineSync and MachineSetSync controller for converting OwnerReferences in mirrored MachineSets, Machines and InfraMachines have correct ownerReferences.

RadekManak commented:
/test e2e-aws-ovn-techpreview
/test unit

damdo commented on pkg/controllers/machinesetsync/machineset_sync_controller.go:
```diff
@@ -270,6 +282,23 @@ func (r *MachineSetSyncReconciler) reconcileMAPIMachineSetToCAPIMachineSet(ctx c
 	newCAPIMachineSet.SetNamespace(r.CAPINamespace)
 	newCAPIMachineSet.Spec.Template.Spec.InfrastructureRef.Namespace = r.CAPINamespace
 
+	convertedCAPIMachineSetOwnerReferences, err := r.convertMAPIMachineSetOwnerReferencesToCAPI(ctx, mapiMachineSet)
```
Could we return something that is also usable later for newCAPIInfraMachineTemplate instead of only for newCAPIMachineSet?

damdo commented on pkg/controllers/machinesetsync/machineset_sync_controller.go:
```diff
@@ -285,6 +314,20 @@ func (r *MachineSetSyncReconciler) reconcileMAPIMachineSetToCAPIMachineSet(ctx c
 	newCAPIInfraMachineTemplate.SetResourceVersion(util.GetResourceVersion(infraMachineTemplate))
 	newCAPIInfraMachineTemplate.SetNamespace(r.CAPINamespace)
 
+	cluster := &capiv1beta1.Cluster{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: r.CAPINamespace, Name: r.Infra.Status.InfrastructureName}, cluster); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get CAPI cluster: %w", err)
+	}
+
+	newCAPIInfraMachineTemplate.SetOwnerReferences([]metav1.OwnerReference{{
+		APIVersion:         capiv1beta1.GroupVersion.String(),
+		Kind:               capiv1beta1.ClusterKind,
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		Controller:         ptr.To(false),
+		BlockOwnerDeletion: ptr.To(true),
+	}})
+
```
See above comment:

> Could we return something that is also usable later for newCAPIInfraMachineTemplate instead of only for newCAPIMachineSet?

damdo commented on pkg/controllers/machinesetsync/machineset_sync_controller.go:
```diff
@@ -336,6 +381,17 @@ func (r *MachineSetSyncReconciler) reconcileCAPIMachineSetToMAPIMachineSet(ctx c
 
+	newMapiMachineSet.OwnerReferences, err = r.convertCAPIMachineSetOwnerReferencesToMAPI(capiMachineSet)
+	if err != nil {
+		logger.Error(err, "Failed to convert Cluster API machineSet owner references to Machine API")
```
```suggestion
		logger.Error(err, "unable to convert Cluster API machine set to Machine API. Cluster API machine set has non-convertible owner references")
```

damdo commented on pkg/controllers/machinesetsync/machineset_sync_controller_test.go:
```diff
@@ -208,6 +226,16 @@ var _ = Describe("With a running MachineSetSync controller", func() {
 					)).Should(Succeed())
 				})
 
+				It("should create MachineSet and MachineTemplate with CAPI Cluster OwnerReference", func() {
```
```suggestion
				It("should create MachineSet and InfraMachineTemplate with CAPI Cluster OwnerReference", func() {
```

damdo commented on pkg/controllers/machinesetsync/machineset_sync_controller_test.go:
```diff
@@ -227,6 +255,19 @@ var _ = Describe("With a running MachineSetSync controller", func() {
 					Expect(k8sClient.Create(ctx, capiMachineSet)).Should(Succeed())
 				})
 
+				It("should update MachineSet and MachineTemplate with CAPI Cluster OwnerReference", func() {
```
```suggestion
				It("should update MachineSet and InfraMachineTemplate with CAPI Cluster OwnerReference", func() {
```

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -236,6 +247,21 @@ func (r *MachineSyncReconciler) reconcileMAPIMachinetoCAPIMachine(ctx context.Co
+
+		var fe *field.Error
+		if errors.As(err, &fe) || errors.Is(err, errUnsuportedOwnerKindForConversion) {
+			logger.Error(err, "Failed to convert Machine API machine owner references to Cluster API")
```
Let's change this to the same msg change as above.

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -464,6 +501,106 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+		// Get the CAPI machineSet with same name as MAPI machineSet
+		if err := r.Get(ctx, key, &capiMachineSet); err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil, fmt.Errorf("cluster API MachineSet not found: %w", err)
```
```suggestion
				return nil, fmt.Errorf("Cluster API machine set not found: %w", err)
```

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -464,6 +501,106 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+			if apierrors.IsNotFound(err) {
+				return nil, fmt.Errorf("cluster API MachineSet not found: %w", err)
+			} else {
+				return nil, fmt.Errorf("error getting MachineSet: %w", err)
```
```suggestion
				return nil, fmt.Errorf("error getting Cluster API machine set: %w", err)
```

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -464,6 +501,106 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+		// Get the MAPI machineSet with same name as CAPI machineSet
+		if err := r.Get(ctx, key, &mapiMachineSet); err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil, fmt.Errorf("machine API MachineSet not found: %w", err)
```
```suggestion
				return nil, fmt.Errorf("Machine API machine set not found: %w", err)
```

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -464,6 +501,106 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+			if apierrors.IsNotFound(err) {
+				return nil, fmt.Errorf("machine API MachineSet not found: %w", err)
+			} else {
+				return nil, fmt.Errorf("error getting MachineSet: %w", err)
```
```suggestion
				return nil, fmt.Errorf("error getting Machine API machine set: %w", err)
```

damdo reviewed (commented):
Thanks for working on this!
Left a bunch of nitpicks but the main logic LGTM

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -497,9 +634,15 @@ func (r *MachineSyncReconciler) fetchCAPIInfraResources(ctx context.Context, cap
+	capiMachine1Copy := capiMachine1.DeepCopy()
+	capiMachine2Copy := capiMachine2.DeepCopy()
+
+	capiMachine1Copy.OwnerReferences = nil
+	capiMachine2Copy.OwnerReferences = nil
+
```
So if these are always nil, how do we pick up the diff there?

RadekManak replied:
This was incorrect.  I think I needed to compare the owner references differently if there could be extra ones that we would not be converting. This was not the way to do It and is not required as the current iteration only allows machine set ownerreference conversion.

damdo commented on pkg/controllers/machinesync/machine_sync_controller_test.go:
```diff
@@ -369,11 +441,35 @@ var _ = Describe("With a running MachineSync Reconciler", func() {
 					Eventually(k.Get(capiInfraMachine)).Should(Succeed(), "should have succeeded getting a CAPI Infra Machine")
 				})
 
+				It("should have OwnerReference from InfraMachine to Machine", func() {
```
```suggestion
				It("should have InfraMachine's OwnerReference on the Machine", func() {
```

JoelSpeed replied:
Other way around now?

damdo commented on pkg/controllers/machinesetsync/machineset_sync_controller.go:
```diff
@@ -362,6 +418,48 @@ func (r *MachineSetSyncReconciler) reconcileCAPIMachineSetToMAPIMachineSet(ctx c
+	}
+
+	// MAPI MachineSet does not have owner references.
+	return []metav1.OwnerReference{}, nil
```
Personally I think the function name correctly captures what we are doing as a whole, but yes, I agree we should add a comment explaining what's happening inside the fuction.

theobarberbany replied:
Yea, I think my naming gripe can be a non-blocking nit:  We're effectively unsetting OwnerReferences, right? Not converting? 

I think _convert_ might be a bit overzealous given we're only functional in one particularly specific scenario...

RadekManak replied:
I renamed the function as it only handles validation now to reflect that.

JoelSpeed commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -79,6 +85,9 @@ var (
 	errUnrecognizedConditionStatus = errors.New("error unrecognized condition status")
+
+	// errUnsuportedOwnerKindForConversion is returned when attempting to convert unsupported ownerReference.
+	errUnsuportedOwnerKindForConversion = errors.New("unsupported owner kind for ownerReference conversion")
```
```suggestion
	errUnsuportedOwnerKindForConversion = errors.New("unsupported owner kind for owner reference conversion")
```

JoelSpeed commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -464,6 +501,106 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+func (r *MachineSyncReconciler) convertMAPIMachineOwnerReferencesToCAPI(ctx context.Context, mapiMachine *machinev1beta1.Machine) ([]metav1.OwnerReference, error) {
+	capiOwnerReferences := []metav1.OwnerReference{}
+
+	if len(mapiMachine.OwnerReferences) > 1 {
+		return nil, field.TooMany(field.NewPath("metadata", "ownerReferences"), len(mapiMachine.OwnerReferences), 1)
+	} else if len(mapiMachine.OwnerReferences) == 1 {
+		ownerRef := mapiMachine.OwnerReferences[0]
+		if ownerRef.Kind != machineSetKind || ownerRef.APIVersion != machinev1beta1.GroupVersion.String() {
+			return nil, field.Invalid(field.NewPath("metadata", "ownerReferences"), mapiMachine.OwnerReferences, errUnsuportedOwnerKindForConversion.Error())
+		}
+	}
```
This seems redundant when we have line 520 no?

JoelSpeed commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -464,6 +501,106 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+func (r *MachineSyncReconciler) convertCAPIMachineOwnerReferencesToMAPI(ctx context.Context, capiMachine *capiv1beta1.Machine) ([]metav1.OwnerReference, error) { //nolint:unused
+	mapiOwnerReferences := []metav1.OwnerReference{}
+
+	if len(capiMachine.OwnerReferences) > 1 {
+		return nil, field.TooMany(field.NewPath("metadata", "ownerReferences"), len(capiMachine.OwnerReferences), 1)
+	} else if len(capiMachine.OwnerReferences) == 1 {
+		ownerRef := capiMachine.OwnerReferences[0]
+		if ownerRef.Kind != machineSetKind || ownerRef.APIVersion != capiv1beta1.GroupVersion.String() {
+			return nil, field.Invalid(field.NewPath("metadata", "ownerReferences"), capiMachine.OwnerReferences, errUnsuportedOwnerKindForConversion.Error())
+		}
+	}
```
Is this redundant considering 570?

JoelSpeed commented on pkg/controllers/machinesync/machine_sync_controller_test.go:
```diff
@@ -99,23 +108,73 @@ var _ = Describe("With a running MachineSync Reconciler", func() {
+		// Create the MAPI MachineSet to have valid owner reference to it
+		mapiMachineSet = mapiMachineSetBuilder.Build()
+		Expect(k8sClient.Create(ctx, mapiMachineSet)).Should(Succeed())
+		ownerReferencesToMapiMachineSet = []metav1.OwnerReference{{
+			APIVersion:         machinev1beta1.GroupVersion.String(),
+			Kind:               machineSetKind,
+			Name:               mapiMachineSet.Name,
+			UID:                mapiMachineSet.UID,
+			Controller:         ptr.To(true),
+			BlockOwnerDeletion: ptr.To(true),
+		},
+		}
```
Maybe this would be neater?
```suggestion
		ownerReferencesToMapiMachineSet = []metav1.OwnerReference{
			{
				APIVersion:         machinev1beta1.GroupVersion.String(),
				Kind:               machineSetKind,
				Name:               mapiMachineSet.Name,
				UID:                mapiMachineSet.UID,
				Controller:         ptr.To(true),
				BlockOwnerDeletion: ptr.To(true),
			},
		}
```

JoelSpeed commented on pkg/controllers/machinesync/machine_sync_controller_test.go:
```diff
@@ -99,23 +108,73 @@ var _ = Describe("With a running MachineSync Reconciler", func() {
+			WithNamespace(capiNamespace.GetName()).WithName("foo-machineset").WithTemplate(capiMachineTemplate).WithClusterName(infrastructureName).
+			Build()
+		Expect(k8sClient.Create(ctx, capiMachineSet)).Should(Succeed())
+		ownerReferencesToCapiMachineSet = []metav1.OwnerReference{{
```
Indentation weird here too

JoelSpeed reviewed (commented):
Does this also cover https://issues.redhat.com/browse/OCPCLOUD-2748? Do we need discussion about the use cases for MachineSet owner references?

damdo commented:
@RadekManak we had a chat with @JoelSpeed about whether this fully covers https://issues.redhat.com/browse/OCPCLOUD-2748 or not.

We think it doesn't fully, but that's fine. We'll come back to https://issues.redhat.com/browse/OCPCLOUD-2748 once we are in 4.20 land and add in/modify what's left to be done.

RadekManak commented:
I think this is ready for re-review. 

[OCPCLOUD-2748](https://issues.redhat.com/browse/OCPCLOUD-2748) Is not fully covered in this PR. 
This PR only handles MAPI/CAPI ownerReferences that are known. Any usecases with third-party owner references should still be explored in that card.

damdo commented:
I just ran this locally and it looks like the happy path is working.
One thing, I noticed we get this log line:
```
E0326 16:24:36.046388   62135 machine_sync_controller.go:360] "unable to convert Machine API machine to Cluster API, unsupported owner reference in conversion" err="metadata.ownerReferences: Invalid value: []v1.OwnerReference{v1.OwnerReference{APIVersion:\"machine.openshift.io/v1\", Kind:\"ControlPlaneMachineSet\", Name:\"cluster\", UID:\"3927d542-7509-46ee-8d98-6367cf6fcac4\", Controller:(*bool)(0x140016cb0ea), BlockOwnerDeletion:(*bool)(0x140016cb0eb)}}: unsupported owner kind for owner reference conversion" controller="MachineSyncController" controllerGroup="machine.openshift.io" controllerKind="Machine" Machine="openshift-machine-api/ddonati-test916-j8fnc-master-2" namespace="openshift-machine-api" name="ddonati-test916-j8fnc-master-2" reconcileID="56d01f34-ce3e-426b-8abd-1c43291abd35"
```

Should we create a special case to not error/log for CPMS Machines, to avoid spamming the logs?
We can add a comment with a TODO once we add handling the conversion for those.

JoelSpeed commented:
> Should we create a special case to not error/log for CPMS Machines, to avoid spamming the logs?
We can add a comment with a TODO once we add handling the conversion for those.

Or, once we've determined that a Machine cannot be converted, we don't requeue? That means not returning the error to controller runtime so we will have to recognise when an error is a conversion error that isn't going to be fixed by requeueing

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -464,6 +499,102 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+	return capiOwnerReferences, nil
+}
+
+// convertCAPIMachineOwnerReferencesToMAPI converts CAPI machine ownerReferences to MAPI ownerReferences.
+//
+//nolint:dupl
+func (r *MachineSyncReconciler) convertCAPIMachineOwnerReferencesToMAPI(ctx context.Context, capiMachine *capiv1beta1.Machine) ([]metav1.OwnerReference, error) { //nolint:unused
```
Is this used somewhere?

RadekManak replied:
This will be called from reconcileCAPIMachinetoMAPIMachine whe implmeneted.

JoelSpeed commented:
I think it's valid to log when there's a conversion error. I don't think it's valid to requeue in that case. If we log the error, it should only happen when there's an update to the Machine right?

damdo commented:
> I think it's valid to log when there's a conversion error. I don't think it's valid to requeue in that case. If we log the error, it should only happen when there's an update to the Machine right?

Ok so let's only log.Error instead of erroring, in case of owner ref conversion errors.

For the CPMS I don't think there is value to us/the user to log.error when converting the CPMS Machines, those are out of scope for this migration so IMO we shouldn't log.error without a real reason.

JoelSpeed commented:
In the future they will be in scope though 🤔

damdo commented:
I don't see the need to log errors just because this might be possible in the future. But I am happy to yield if you think it is important.

damdo commented:
> Ok so let's only log.Error instead of erroring in any case of owner ref conversion errors.

@RadekManak as a matter of fact, I think we could do the owner reference check / conversion (`convertMAPIMachineOwnerReferencesToCAPI()`) before doing the whole `convertMAPIToCAPIMachine()`, so we bail early if we shouldn't/aren't convert the owner refs. Same for the MachineSet. WDYT?

damdo commented on pkg/controllers/machinesetsync/machineset_sync_controller.go:
```diff
@@ -270,6 +280,42 @@ func (r *MachineSetSyncReconciler) reconcileMAPIMachineSetToCAPIMachineSet(ctx c
 	newCAPIMachineSet.SetNamespace(r.CAPINamespace)
 	newCAPIMachineSet.Spec.Template.Spec.InfrastructureRef.Namespace = r.CAPINamespace
 
+	err = r.validateMAPIMachineSetOwnerReferences(mapiMachineSet)
+	if err != nil {
```
nit, would be cleaner to do inline `if err:= func(); err !=nil {` here

damdo commented on pkg/controllers/machinesetsync/machineset_sync_controller.go:
```diff
@@ -336,6 +380,19 @@ func (r *MachineSetSyncReconciler) reconcileCAPIMachineSetToMAPIMachineSet(ctx c
 	// The conversion does not set a resource version, so we must copy it over
 	newMapiMachineSet.SetResourceVersion(util.GetResourceVersion(mapiMachineSet))
 
+	err = r.validateCAPIMachineSetOwnerReferences(capiMachineSet)
+	if err != nil {
```
Let's inline the error here too

RadekManak commented:
Moved the owner references to be checked first.

I added new error for control plane machines. Instead of printing the generic error, It now info logs that the conversion is currently not supported. The condition will have the new error message. 

This should prevent any confusion as to why the conversion is failing and still keep the signal as to why the machine is not being mirrored into CAPI.

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -214,9 +227,32 @@ func (r *MachineSyncReconciler) reconcileCAPIMachinetoMAPIMachine(ctx context.Co
+
+		var fe *field.Error
+		if errors.As(err, &fe) || errors.Is(err, errUnsuportedOwnerKindForConversion) {
+			if fe.Detail == errUnsuportedOwnerKindForConversion.Error() {
```
Should this be `errUnsupportedCPMSOwnedMachineConversion`?

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -214,9 +227,32 @@ func (r *MachineSyncReconciler) reconcileCAPIMachinetoMAPIMachine(ctx context.Co
 
+	newCAPIOwnerReferences, err := r.convertMAPIMachineOwnerReferencesToCAPI(ctx, mapiMachine)
+	if err != nil {
+		if condErr := r.applySynchronizedConditionWithPatch(ctx, mapiMachine, corev1.ConditionFalse, reasonFailedToConvertMAPIMachineToCAPI, err.Error(), nil); condErr != nil {
+			return ctrl.Result{}, utilerrors.NewAggregate([]error{err, condErr})
+		}
```
This here causes a loop as it triggers re-reconciles also in cases where we don't want to, like in the `errUnsuportedOwnerKindForConversion` cases below.

damdo commented on pkg/controllers/machinesync/machine_sync_controller.go:
```diff
@@ -214,9 +227,32 @@ func (r *MachineSyncReconciler) reconcileCAPIMachinetoMAPIMachine(ctx context.Co
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, fmt.Errorf("failed to convert Machine API machine owner references to Cluster API: %w", err)
```
I don't think we should ever error in case of failed owner reference conversion.
As we decided here.
https://github.com/openshift/cluster-capi-operator/pull/271#issuecomment-2754897624

damdo replied:
This will otherwise cause infinite reconcile loops when, for example, MAPI MachineSets are owned by  more than one owner reference.

See for exampl case 
```
	if len(mapiMachine.OwnerReferences) > 1 {
```
in `convertMAPIMachineOwnerReferencesToCAPI()`.

RadekManak replied:
This should now only retry when
	if err := r.Get(ctx, key, &capiMachineSet); err != nil {
fails.

The terminal field errors are handled above.

damdo reviewed (commented):
Thanks @RadekManak for addressing the feedback.
From what I can see all the feedback that @theobarberbany @JoelSpeed and I had was addressed

I also tested the changes locally and things were flowing correctly for the happy cases I rehearsed.

As I was mentioning to @theobarberbany on #265 we could probably to with a bit more in-depth testing especially on the various error paths, but we may want to coordinate and discuss that effort in combination with some coverage data, to help us better understand how to tackle that.

/approve
/lgtm

damdo commented:
/test e2e-vsphere-capi-techpreview
